feat(diff): cancel linematch if ctrl-c is pressed while it's running

jwhite510 opened the pull request:
@lewis6991 
In the case of running a linematch algorithm with a really large diff hunk that takes forever, the algorithm will hang and the user must stop the neovim process. With this pull request, I added an ability for the running algorithm to respect the ctrl + c press, similar to when you have a word completion (ctrl-p / ctrl-n) huge word list coming from many open buffers or large files, it will hang until you press ctrl + c to cancel it.

I'm not sure if i handled the capture of the ctrl-c the best way, but it seems to work. so some input on that would be appreciated if anyone knows a better way

jwhite510 committed:
feat(diff): cancel linematch if ctrl-c is pressed while running to prevent hang

justinmk commented:
This is definitely wanted, thanks! Typically the way to check for ctrl-c is one of the breakcheck functions https://github.com/neovim/neovim/blob/b923fcbaf06276051372c17177a9970c4adc8e3d/src/nvim/os/input.c#L210-L228

E.g. if this is in a very hot loop, use `veryfast_breakcheck`. Does that not work?

justinmk commented:
@jwhite510 can you rebase this? any thoughts on my comment above?

jwhite510 commented:
@justinmk 
rebased, and now using the the veryfast_breakcheck, that's exactly what I was looking for.

justinmk commented on src/nvim/linematch.c:
```diff
@@ -353,24 +362,46 @@ size_t linematch_nbuffers(const char **diff_blk, const int *diff_len, const size
-    (*decisions)[i] = (*decisions)[n_optimal - 1 - i];
-    (*decisions)[n_optimal - 1 - i] = tmp;
+  if (canceled == true) { // if canceled, compare the lines guessing that the new lines are at the
+                          // bottom of the longer text
```
would it simplify things we just bailed out of diff-mode entirely? I don't think we need to bother trying to construct partial "guessed" result.

justinmk commented:
Needs a rebase, and a test.

jwhite510 committed:
add an incomplete test

jwhite510 commented:
@justinmk 
> Needs a rebase, and a test. And hopefully the code will be much simpler after removing the attempt to "guess" when the user cancels.

I removed the simplified guessed result, so the error handling is now moved outside linematch_nbuffers. I stored the failed state in diff_T.is_linematched, which is now an integer. 0: not yet ran, 1: success, -1: canceled. I'm not sure in the lua test how to simulate the ctrl-c press of the user to cancel it while it is still running.

justinmk commented on test/functional/ui/linematch_spec.lua:
```diff
@@ -1084,6 +1084,38 @@ something
+    before_each(function()
+      local f1 = [[
+a
+a
```
can use `t.dedent()` to make this look nicer :)

justinmk commented:
sorry for missing this, if you can rebase and CI looks good let's merge it.
